channels/reducer.js

```diff
@@ -117,6 +98,49 @@ function createReducer (config) {
           [id]: merge(state.records[id], action.payload)
         }
       }
+    },
+    [createPrimaryTrackFromFile]: (state, action) => {
+      const { file, parentChannelId, attrs = {} } = action.payload
+
+      const effectCreator = (sampleId) => {
+        const channelId = uuid()
+        const clipId = uuid()
+
+        return Effects.batch([
+          Effects.constant(createClip({ id: clipId, sampleId })),
+          Effects.constant(createChannel(merge({
+            id: channelId,
+            type: CHANNEL_TYPE_PRIMARY_TRACK,
+            clipIds: [clipId] // TODO(FUTURE): maybe setClipChannel?
+          }, attrs))),
+          Effects.constant(setChannelParent({ parentChannelId, channelId }))
+        ])
+      }
+
+      return loop(state, Effects.constant(createSample({ file, effectCreator })))
+    },
+    [swapPrimaryTracks]: (state, action) => {
+      const { sourceId, targetId } = action.payload
+      if (sourceId === targetId) { return state }
+
+      // TODO(FUTURE): also update both associated transitions to same endBeat as track
+      // TODO(FUTURE): anything we need to do with track or transition length?
+      const source = clone(state.records[sourceId])
+      const target = clone(state.records[targetId])
+
+      return {
+        ...state,
+        dirty: [...state.dirty, sourceId, targetId],
+        records: {
+          ...state.records,
+          [sourceId]: merge(state.records[sourceId], {
+            startBeat: target.startBeat
+          }),
+          [targetId]: merge(state.records[targetId], {
+            startBeat: source.startBeat
+          })
```

Comment:
i couldnt loop `updateChannel` here because i needed the store to be updated for the next `swapPrimaryTracks` effect. hopefully wont be runnign into this often, and if i am, should figure out a good pattern